### PR TITLE
Update first_occurance.py

### PR DESCRIPTION
**Pull request by tratneshsingh**
The earlier code fails to output the right index for following input: (Note that the first value '1' is repeated)
array = [1, 1, 2, 3, 3, 3, 3, 4, 4, 4, 4, 5, 6, 6, 6]
firstOccurance(array, 1)

### CODE REVIEW

File: search/first_occurance.py [hunk 1]
```diff
@@ -4,42 +4,34 @@
...
-    print("found first: ", query, " in index:", firstOccurance(array, query))
-    print("-----SEARCH-----")
-    query = 9
-    print("found first: ", query, " in index:", firstOccurance(array, query))
-
-if __name__ == "__main__":
-    main()
+    lo, hi = 0, len(array)
+    while lo < hi:
+        mid = lo + (hi - lo) // 2
```
Comment (christianbender) [inline search/first_occurance.py]: Here you can write ```(hi + lo) // 2```

Comment (christianbender) [inline search/first_occurance.py]: Write ```mid = (hi + lo) // 2```

File: search/first_occurance.py [hunk 2]
```diff
@@ -4,42 +4,34 @@
...
-    print("-----SEARCH-----")
-    query = -1
-    print("found first: ", query, " in index:", firstOccurance(array, query))
-    print("-----SEARCH-----")
-    query = 9
-    print("found first: ", query, " in index:", firstOccurance(array, query))
-
-if __name__ == "__main__":
-    main()
+    lo, hi = 0, len(array)
```
Comment (christianbender) [inline search/first_occurance.py]: Instead of ```len(array)``` write ```len(array) -1```

File: search/first_occurance.py [hunk 3]
```diff
@@ -4,42 +4,34 @@
...
+    while lo < hi:
+        mid = lo + (hi - lo) // 2
+        #print("lo: ", lo, " hi: ", hi, " mid: ", mid)
+        if array[mid] < query: lo = mid + 1
```
Comment (christianbender) [inline search/first_occurance.py]: Please write ```lo = mid +1``` in the next line. According to PEP8

File: search/first_occurance.py [hunk 4]
```diff
@@ -4,42 +4,34 @@
...
+        else: hi = mid
```
Comment (christianbender) [inline search/first_occurance.py]: Same thing as above.

File: search/first_occurance.py [hunk 5]
```diff
@@ -7,16 +7,19 @@
 import unittest
 
 def firstOccurance(array, query):
-    lo, hi = 0, len(array)
-    while lo < hi:
-        mid = lo + (hi - lo) // 2
+    lo, hi = 0, len(array) - 1
+    while lo <= hi:
+        mid = (hi + lo) // 2
```
Comment (keon) [inline search/first_occurance.py]: this can cause overflow.
the previous version was right.

File: search/first_occurance.py [hunk 6]
```diff
@@ -7,16 +7,19 @@
...
         #print("lo: ", lo, " hi: ", hi, " mid: ", mid)
-        if array[mid] < query: lo = mid + 1
-        else: hi = mid
-    if lo < len(array) and array[lo] == query:
+        if lo == hi:
```
Comment (keon) [inline search/first_occurance.py]: I don't see how this is more preferable than the previous version.

File: search/first_occurance.py (hunk 5 above)
Comment (tratneshsingh) [reply search/first_occurance.py]: Restored the previous one.

Comment (goswami-rahul) [reply search/first_occurance.py]: @keon I think there is no Integer overflow in Python. Or is there?

File: search/first_occurance.py (hunk 2 above)
Comment (tratneshsingh) [reply search/first_occurance.py]: Done

File: search/first_occurance.py (hunk 1 above)
Comment (tratneshsingh) [reply search/first_occurance.py]: Done

Comment (tratneshsingh) [reply search/first_occurance.py]: Done

File: search/first_occurance.py (hunk 4 above)
Comment (tratneshsingh) [reply search/first_occurance.py]: Done

File: search/first_occurance.py (hunk 5 above)
Comment (keon) [reply search/first_occurance.py]: @goswami-rahul you're right.
but still good to know for other langs tho.

**Pull request merged**